File: tests/test_relu.py
# tests/test_relu.py

import torch
import torch.nn as nn

from layers.ReLU import ManualReLU


RMSE_LIMIT = 1e-3
ATOL = 1e-6
RTOL = 1e-5


def calculate_rmse(reference, prediction):
    """
    Calcula la raíz del error cuadrático medio.

    RMSE = sqrt(mean((reference - prediction)^2))
    """

    if reference.shape != prediction.shape:
        raise ValueError(
            "No se puede calcular el RMSE porque las formas "
            f"son diferentes: {reference.shape} y {prediction.shape}"
        )

    squared_error = (
        reference - prediction
    ) ** 2

    mean_squared_error = torch.mean(
        squared_error
    )

    rmse = torch.sqrt(
        mean_squared_error
    )

    return rmse.item()


def calculate_max_absolute_error(reference, prediction):
    """
    Calcula la diferencia absoluta máxima
    entre ambas salidas.
    """

    if reference.shape != prediction.shape:
        raise ValueError(
            "No se puede calcular el error máximo porque "
            f"las formas son diferentes: "
            f"{reference.shape} y {prediction.shape}"
        )

    absolute_error = torch.abs(
        reference - prediction
    )

    return torch.max(
        absolute_error
    ).item()


def test_manual_relu(shape):
    """
    Compara ManualReLU contra torch.nn.ReLU
    utilizando exactamente la misma entrada.

    También calcula:

        - Error absoluto máximo.
        - RMSE.
        - Coincidencia mediante torch.allclose.

    Parámetro:
        shape: forma de la entrada.
    """

    # ---------------------------------------------------------
    # 1. Reproducibilidad
    # ---------------------------------------------------------

    torch.manual_seed(5)

    # ---------------------------------------------------------
    # 2. Crear las implementaciones
    # ---------------------------------------------------------

    pytorch_relu = nn.ReLU()

    manual_relu = ManualReLU()

    # ---------------------------------------------------------
    # 3. Crear la entrada compartida
    # ---------------------------------------------------------

    x_torch = torch.randn(
        *shape,
        dtype=torch.float32,
    )

    x_manual = x_torch.tolist()

    # ---------------------------------------------------------
    # 4. Ejecutar PyTorch
    # ---------------------------------------------------------

    with torch.no_grad():
        y_pytorch = pytorch_relu(
            x_torch
        )

    # ---------------------------------------------------------
    # 5. Ejecutar ManualReLU
    # ---------------------------------------------------------

    y_manual_list = manual_relu(
        x_manual
    )

    y_manual = torch.tensor(
        y_manual_list,
        dtype=torch.float32,
    )

    # ---------------------------------------------------------
    # 6. Comparar formas
    # ---------------------------------------------------------

    same_shape = (
        y_pytorch.shape
        == y_manual.shape
    )

    # ---------------------------------------------------------
    # 7. Calcular métricas
    # ---------------------------------------------------------

    rmse = calculate_rmse(
        y_pytorch,
        y_manual,
    )

    max_error = calculate_max_absolute_error(
        y_pytorch,
        y_manual,
    )

    # ---------------------------------------------------------
    # 8. Comparar valores
    # ---------------------------------------------------------

    values_match = torch.allclose(
        y_pytorch,
        y_manual,
        atol=ATOL,
        rtol=RTOL,
    )

    rmse_is_valid = (
        rmse <= RMSE_LIMIT
    )

    # ---------------------------------------------------------
    # 9. Mostrar resultados
    # ---------------------------------------------------------

    print("=" * 60)
    print("VALIDACIÓN DE ManualReLU")
    print("=" * 60)

    print(f"Forma de entrada:       {shape}")
    print(f"Cantidad de elementos:  {x_torch.numel()}")
    print(f"Forma PyTorch:          {tuple(y_pytorch.shape)}")
    print(f"Forma ManualReLU:       {tuple(y_manual.shape)}")
    print(f"Misma forma:            {same_shape}")
    print(f"Error absoluto máximo:  {max_error:.10e}")
    print(f"RMSE:                   {rmse:.10e}")
    print(f"Límite RMSE:            {RMSE_LIMIT:.10e}")
    print(f"torch.allclose:          {values_match}")
    print(f"RMSE válido:            {rmse_is_valid}")

    if (
        same_shape
        and values_match
        and rmse_is_valid
    ):
        print(
            "Resultado:               APROBADO"
        )

    else:
        print(
            "Resultado:               NO APROBADO"
        )

    print("=" * 60)
    print()

    # ---------------------------------------------------------
    # 10. Verificación automática
    # ---------------------------------------------------------

    assert same_shape, (
        "La forma de salida de ManualReLU "
        "no coincide con PyTorch"
    )

    assert values_match, (
        "Los valores de ManualReLU "
        "no coinciden con torch.nn.ReLU"
    )

    assert rmse_is_valid, (
        f"El RMSE {rmse} supera el límite "
        f"permitido de {RMSE_LIMIT}"
    )

    return {
        "x": x_torch,
        "y_pytorch": y_pytorch,
        "y_manual": y_manual,
        "max_error": max_error,
        "rmse": rmse,
        "same_shape": same_shape,
        "values_match": values_match,
    }


def test_known_values():
    """
    Comprueba ManualReLU con valores conocidos.
    """

    manual_relu = ManualReLU()

    x_manual = [
        [-3.0, -1.0, 0.0],
        [1.0, 2.5, -4.5],
    ]

    expected_output = [
        [0.0, 0.0, 0.0],
        [1.0, 2.5, 0.0],
    ]

    output_manual = manual_relu(
        x_manual
    )

    assert output_manual == expected_output


def test_input_is_not_modified():
    """
    Comprueba que ManualReLU no modifique
    la lista de entrada original.
    """

    manual_relu = ManualReLU()

    x_manual = [
        [-2.0, 0.0, 3.0],
        [4.0, -5.0, 6.0],
    ]

    original_input = [
        row.copy()
        for row in x_manual
    ]

    manual_relu(
        x_manual
    )

    assert x_manual == original_input


def test_invalid_non_list_input():
    """
    Comprueba que ManualReLU rechace
    una entrada que no sea una lista.
    """

    manual_relu = ManualReLU()

    try:
        manual_relu(
            4.0
        )

    except TypeError:
        pass

    else:
        raise AssertionError(
            "ManualReLU debía lanzar TypeError "
            "para una entrada que no fuera una lista"
        )


def test_empty_input():
    """
    Comprueba que ManualReLU rechace
    una lista vacía.
    """

    manual_relu = ManualReLU()

    try:
        manual_relu(
            []
        )

    except ValueError:
        pass

    else:
        raise AssertionError(
            "ManualReLU debía lanzar ValueError "
            "para una lista vacía"
        )


def test_nested_empty_list():
    """
    Comprueba que ManualReLU rechace
    listas vacías anidadas.
    """

    manual_relu = ManualReLU()

    try:
        manual_relu(
            [
                [1.0, 2.0],
                [],
            ]
        )

    except ValueError:
        pass

    else:
        raise AssertionError(
            "ManualReLU debía lanzar ValueError "
            "para una lista vacía anidada"
        )


def test_invalid_element():
    """
    Comprueba que ManualReLU rechace
    elementos no numéricos.
    """

    manual_relu = ManualReLU()

    try:
        manual_relu(
            [
                [1.0, "dato inválido"],
            ]
        )

    except TypeError:
        pass

    else:
        raise AssertionError(
            "ManualReLU debía lanzar TypeError "
            "para elementos no numéricos"
        )


def main():
    """
    Ejecuta el banco completo de pruebas
    de ManualReLU.
    """

    test_shapes = [
        # Vector simple
        (10,),

        # Salida de una capa Linear
        (8, 16),

        # Entrada similar a un batch de MNIST
        (4, 1, 28, 28),

        # Activación intermedia similar a LeNet-5
        (4, 6, 24, 24),

        # Entrada de una imagen ImageNet
        (1, 3, 224, 224),

        # Activación interna similar a MobileNetV2
        (1, 32, 112, 112),
    ]

    results = []

    print()
    print("#" * 60)
    print("BANCO DE PRUEBAS DE ManualReLU")
    print("#" * 60)
    print()

    for shape in test_shapes:
        result = test_manual_relu(
            shape=shape,
        )

        results.append(
            result
        )

    # ---------------------------------------------------------
    # Pruebas adicionales
    # ---------------------------------------------------------

    test_known_values()

    test_input_is_not_modified()

    test_invalid_non_list_input()

    test_empty_input()

    test_nested_empty_list()

    test_invalid_element()

    # ---------------------------------------------------------
    # Resumen final
    # ---------------------------------------------------------

    print("#" * 60)
    print("RESUMEN GENERAL")
    print("#" * 60)

    for index, result in enumerate(
        results,
        start=1,
    ):
        print(
            f"Prueba {index}: "
            f"shape={tuple(result['x'].shape)}, "
            f"RMSE={result['rmse']:.10e}, "
            f"error máximo={result['max_error']:.10e}"
        )

    print()
    print(
        "Pruebas de valores conocidos:     APROBADAS"
    )

    print(
        "Prueba de entrada no modificada:  APROBADA"
    )

    print(
        "Pruebas de validación de entrada: APROBADAS"
    )

    print()
    print(
        "TODAS LAS PRUEBAS DE ManualReLU "
        "FUERON SUPERADAS."
    )

    print("#" * 60)


if __name__ == "__main__":
    main()

File: cpp/tests/test_relu.cpp
#include "ReLU.hpp"

#include <cmath>
#include <iostream>
#include <random>
#include <vector>
#include <string>
#include <iomanip>

constexpr float RMSE_LIMIT = 1e-3f;
constexpr float ATOL = 1e-6f;
constexpr float RTOL = 1e-5f;


/*
 * Calcula:
 *
 * RMSE = sqrt(mean((reference - prediction)^2))
 */
float calculate_rmse(
    const std::vector<float>& reference,
    const std::vector<float>& prediction
)
{
    if (reference.size() != prediction.size()) {
        throw std::runtime_error(
            "No se puede calcular RMSE: tamaños diferentes."
        );
    }

    double squared_error_sum = 0.0;

    for (std::size_t i = 0; i < reference.size(); ++i) {
        const double error =
            static_cast<double>(reference[i]) -
            static_cast<double>(prediction[i]);

        squared_error_sum += error * error;
    }

    const double mean_squared_error =
        squared_error_sum /
        static_cast<double>(reference.size());

    return static_cast<float>(
        std::sqrt(mean_squared_error)
    );
}


/*
 * Calcula el error absoluto máximo.
 */
float calculate_max_absolute_error(
    const std::vector<float>& reference,
    const std::vector<float>& prediction
)
{
    if (reference.size() != prediction.size()) {
        throw std::runtime_error(
            "No se puede calcular error máximo: tamaños diferentes."
        );
    }

    float max_error = 0.0f;

    for (std::size_t i = 0; i < reference.size(); ++i) {
        const float error = std::fabs(
            reference[i] - prediction[i]
        );

        if (error > max_error) {
            max_error = error;
        }
    }

    return max_error;
}


/*
 * Equivalente aproximado de torch.allclose().
 *
 * |a - b| <= atol + rtol * |b|
 */
bool allclose(
    const std::vector<float>& reference,
    const std::vector<float>& prediction
)
{
    if (reference.size() != prediction.size()) {
        return false;
    }

    for (std::size_t i = 0; i < reference.size(); ++i) {
        const float difference = std::fabs(
            reference[i] - prediction[i]
        );

        const float tolerance =
            ATOL + RTOL * std::fabs(reference[i]);

        if (difference > tolerance) {
            return false;
        }
    }

    return true;
}


/*
 * Convierte una forma multidimensional en
 * cantidad total de elementos.
 *
 * Ejemplo:
 *
 * 4 x 6 x 24 x 24 = 13824 elementos
 */
std::size_t calculate_num_elements(
    const std::vector<int>& shape
)
{
    std::size_t total = 1;

    for (int dimension : shape) {
        total *= static_cast<std::size_t>(dimension);
    }

    return total;
}


/*
 * Convierte la forma a texto.
 */
std::string shape_to_string(
    const std::vector<int>& shape
)
{
    std::string text = "(";

    for (std::size_t i = 0; i < shape.size(); ++i) {
        text += std::to_string(shape[i]);

        if (i + 1 < shape.size()) {
            text += ", ";
        }
    }

    text += ")";

    return text;
}


/*
 * Prueba principal de ReLU.
 */
bool test_relu(
    const std::vector<int>& shape
)
{
    const std::size_t size =
        calculate_num_elements(shape);

    /*
     * Generador reproducible.
     *
     * Equivalente conceptual a:
     *
     * torch.manual_seed(5)
     */
    std::mt19937 generator(5);

    std::normal_distribution<float> distribution(
        0.0f,
        1.0f
    );

    /*
     * Entrada compartida.
     */
    std::vector<float> input(size);

    for (float& value : input) {
        value = distribution(generator);
    }

    /*
     * Referencia.
     *
     * Este bloque representa el comportamiento
     * esperado de ReLU.
     */
    std::vector<float> reference(size);

    for (std::size_t i = 0; i < size; ++i) {
        reference[i] =
            input[i] > 0.0f
                ? input[i]
                : 0.0f;
    }

    /*
     * Ejecutar nuestra implementación C++.
     */
    std::vector<float> output(size);

    relu_forward(
        input.data(),
        output.data(),
        static_cast<int>(size)
    );

    /*
     * Calcular métricas.
     */
    const float rmse = calculate_rmse(
        reference,
        output
    );

    const float max_error =
        calculate_max_absolute_error(
            reference,
            output
        );

    const bool values_match = allclose(
        reference,
        output
    );

    const bool rmse_is_valid =
        rmse <= RMSE_LIMIT;

    /*
     * Mostrar resultados.
     */
    std::cout
        << "============================================================\n";

    std::cout
        << "VALIDACION DE ReLU C++\n";

    std::cout
        << "============================================================\n";

    std::cout
        << "Forma de entrada:       "
        << shape_to_string(shape)
        << '\n';

    std::cout
        << "Cantidad de elementos:  "
        << size
        << '\n';

    std::cout
        << std::scientific
        << std::setprecision(10);

    std::cout
        << "Error absoluto maximo:  "
        << max_error
        << '\n';

    std::cout
        << "RMSE:                   "
        << rmse
        << '\n';

    std::cout
        << "Limite RMSE:            "
        << RMSE_LIMIT
        << '\n';

    std::cout
        << std::boolalpha;

    std::cout
        << "allclose:               "
        << values_match
        << '\n';

    std::cout
        << "RMSE valido:            "
        << rmse_is_valid
        << '\n';

    const bool passed =
        values_match &&
        rmse_is_valid;

    std::cout
        << "Resultado:               "
        << (passed ? "APROBADO" : "NO APROBADO")
        << '\n';

    std::cout
        << "============================================================\n\n";

    return passed;
}


/*
 * Prueba con valores conocidos.
 */
bool test_known_values()
{
    const std::vector<float> input = {
        -3.0f,
        -1.0f,
        0.0f,
        1.0f,
        2.5f,
        -4.5f
    };

    const std::vector<float> expected = {
        0.0f,
        0.0f,
        0.0f,
        1.0f,
        2.5f,
        0.0f
    };

    std::vector<float> output(
        input.size()
    );

    relu_forward(
        input.data(),
        output.data(),
        static_cast<int>(input.size())
    );

    return allclose(
        expected,
        output
    );
}


int main()
{
    const std::vector<std::vector<int>> test_shapes = {

        // Vector simple
        {10},

        // Salida de Linear
        {8, 16},

        // Batch similar a MNIST
        {4, 1, 28, 28},

        // Activacion similar a LeNet-5
        {4, 6, 24, 24},

        // Imagen ImageNet
        {1, 3, 224, 224},

        // Activacion similar a MobileNetV2
        {1, 32, 112, 112}
    };

    std::cout << '\n';

    std::cout
        << "############################################################\n";

    std::cout
        << "BANCO DE PRUEBAS DE ReLU C++\n";

    std::cout
        << "############################################################\n\n";


    bool all_tests_passed = true;

    int test_number = 1;

    for (const auto& shape : test_shapes) {

        const bool passed =
            test_relu(shape);

        if (!passed) {
            all_tests_passed = false;
        }

        std::cout
            << "Prueba "
            << test_number
            << ": shape="
            << shape_to_string(shape)
            << " -> "
            << (passed ? "APROBADA" : "NO APROBADA")
            << '\n';

        ++test_number;
    }


    /*
     * Valores conocidos.
     */
    const bool known_values_passed =
        test_known_values();

    if (!known_values_passed) {
        all_tests_passed = false;
    }


    std::cout << '\n';

    std::cout
        << "############################################################\n";

    std::cout
        << "RESUMEN GENERAL\n";

    std::cout
        << "############################################################\n";

    std::cout
        << "Prueba de valores conocidos: "
        << (
            known_values_passed
            ? "APROBADA"
            : "NO APROBADA"
        )
        << '\n';


    if (all_tests_passed) {

        std::cout << '\n';

        std::cout
            << "TODAS LAS PRUEBAS DE ReLU C++ "
            << "FUERON SUPERADAS.\n";

        std::cout
            << "############################################################\n";

        return 0;
    }


    std::cout << '\n';

    std::cout
        << "EXISTEN PRUEBAS NO SUPERADAS.\n";

    std::cout
        << "############################################################\n";

    return 1;
}